### ISSUE TITLE
Issue 003: `seisai-engine` の直接依存と feature extra を pyproject に明示する

### ISSUE THREAD

**Issue by ozw4** [ozw4/seisai#227]
# Issue 003: `seisai-engine` の直接依存と feature extra を pyproject に明示する

## 背景

`packages/seisai-engine/pyproject.toml` の現在の依存は、概ね次の通りです。

```toml
dependencies = [
  "numpy",
  "scipy",
  "torch",
  "tqdm",
  "mlflow>=2.0",
  "seisai-dataset>=0.1.0",
  "seisai-transforms>=0.1.0",
  "seisai-models>=0.1.0",
  "seisai-utils>=0.1.0",
]
```

しかし、`packages/seisai-engine/src/seisai_engine` 配下の実コードは、以下も直接 import しています。

- `yaml` / `PyYAML`
  - `seisai_engine/pipelines/common/train_skeleton_tracking.py`
- `segyio`
  - `seisai_engine/pipelines/fbpick/coarse/build_dataset.py`
  - `seisai_engine/pipelines/fbpick/fine/infer.py`
  - `seisai_engine/pipelines/psn/noise_dataset.py`
- `seisai_pick`
  - `seisai_engine/loss/trend_prior_loss.py`
  - `seisai_engine/postprocess/trend_prior_op.py`
  - `seisai_engine/pipelines/fbpick/export/grstat.py`
  - `seisai_engine/pipelines/fbpick/physics/*.py`
- `matplotlib`
  - `seisai_engine/pipelines/fbpick/coarse/eval.py`
  - `seisai_engine/pipelines/fbpick/coarse/qc.py`
  - `seisai_engine/viewer/fbpick.py`
- `timm`
  - `seisai_engine/optim.py` の `build_optimizer()` 内で、optimizer section が指定された場合だけ lazy import

現状は、これらの一部が transitive dependency によって偶然入ることを前提にしています。単独 install や依存グラフ変更時に壊れやすいです。

## 依存方針

この issue では、**通常 engine 利用で必要な直接依存は required に入れ、plot/backend など機能限定の依存だけ extras に分ける**方針にしてください。

- required にするもの:
  - 軽量・通常利用・直接 import の依存
  - engine の主要 pipeline で必要な内部 package
  - SEG-Y pipeline が engine の主要機能として扱われている場合の `segyio`
- optional extras に分けるもの:
  - visualization / QC だけで使う `matplotlib`
  - non-default optimizer backend だけで使う `timm`

この issue では `mlflow>=2.0` の required / optional 方針は変更しないでください。現状の README では tracking を使わない場合でも `mlflow` は依存に含まれる、と説明されているため、optional 化は別 issue で扱います。

## 目的

`seisai-engine` が直接使う runtime dependency を `pyproject.toml` に明示しつつ、`matplotlib` や `timm` のような機能限定依存は extras と lazy import で扱える状態にします。

## 作業内容

1. `packages/seisai-engine/pyproject.toml` の `[project].dependencies` を更新してください。
   - 維持する依存:
     - `numpy`
     - `scipy`
     - `torch`
     - `tqdm`
     - `mlflow>=2.0`
     - `seisai-dataset>=0.1.0`
     - `seisai-transforms>=0.1.0`
     - `seisai-models>=0.1.0`
     - `seisai-utils>=0.1.0`
   - required dependency として追加する依存:
     - `PyYAML`
     - `seisai-pick>=0.1.0`
     - `segyio`

   `segyio` は `seisai-dataset` から transitive に入る可能性がありますが、`seisai-engine` 自身が直接 import しているため、ここで明示してください。

2. `matplotlib` を visualization / QC 用 extra に分けてください。
   - 例:

```toml
[project.optional-dependencies]
viz = ["matplotlib"]
```

   - `seisai_engine.viewer.*` や `fbpick.coarse.qc/eval` など、`matplotlib` が必要な module では、依存不足時に `pip install seisai-engine[viz]` を案内してください。
   - 可能なら `matplotlib` import を関数内に遅延させ、無関係な engine import が `matplotlib` 不足で失敗しないようにしてください。

3. `timm` を optimizer backend 用 extra に分けてください。
   - 例:

```toml
[project.optional-dependencies]
optim = ["timm"]
```

   - `seisai_engine.optim.build_optimizer()` で `optimizer` section が指定され、`timm` が必要になった時だけ import してください。
   - `timm` がない場合は `pip install seisai-engine[optim]` を案内する明確な error にしてください。
   - default path、つまり `optimizer` section がない場合の `torch.optim.AdamW` は `timm` なしで動く状態を維持してください。

4. `packages/seisai-engine/README.md` の Requirements を更新してください。
   - required dependencies
   - optional extras
   - `viz` extra が必要な module / CLI
   - `optim` extra が必要な optimizer config
   - `mlflow` は現状 required のまま維持すること

5. import smoke test を追加してください。
   - 例: `packages/seisai-engine/tests/test_import_dependencies.py`
   - base install で少なくとも以下を import してください。

```python
import seisai_engine
import seisai_engine.optim
import seisai_engine.loss.trend_prior_loss
import seisai_engine.postprocess.trend_prior_op
import seisai_engine.pipelines.common.train_skeleton_tracking
import seisai_engine.pipelines.fbpick.physics.confidence
```

   - `[viz]` extra がある環境では以下も import してください。

```python
import seisai_engine.viewer.fbpick
import seisai_engine.pipelines.fbpick.coarse.qc
```

## 受け入れ条件

- `PyYAML`, `seisai-pick>=0.1.0`, `segyio` が `seisai-engine` の required dependency として明示されている。
- `matplotlib` は required dependency ではなく、visualization / QC 用 optional extra として扱われている。
- `timm` は required dependency ではなく、optimizer backend 用 optional extra として扱われている。
- `build_optimizer()` の default path が `timm` なしで動く。
- README の Requirements が pyproject と矛盾していない。
- import smoke test が追加されている。
- 全 monorepo package を editable install した状態で、`pytest --collect-only -m "not e2e and not integration"` が dependency 不足で失敗しない。

## 検証コマンド例

```bash
python -m pip install -U pip
for p in seisai-utils seisai-pick seisai-transforms seisai-dataset seisai-models seisai-engine; do
  python -m pip install -e "packages/$p"
done
python -m pip check

python - <<'PY'
import seisai_engine
import seisai_engine.optim
import seisai_engine.loss.trend_prior_loss
import seisai_engine.postprocess.trend_prior_op
import seisai_engine.pipelines.common.train_skeleton_tracking
import seisai_engine.pipelines.fbpick.physics.confidence
print('seisai-engine base import smoke OK')
PY

python -m pip install -e 'packages/seisai-engine[viz,optim]'
python -m pip check
python - <<'PY'
import seisai_engine.viewer.fbpick
import seisai_engine.pipelines.fbpick.coarse.qc
print('seisai-engine optional import smoke OK')
PY

python -m pytest --collect-only -m "not e2e and not integration" -q
python -m pytest packages/seisai-engine/tests -q
```

## 非目標

- `mlflow` の optional 化はこの issue の範囲外です。
- engine pipeline の学習・推論ロジック変更は不要です。
- `segyio` の optional 化はこの issue では行わないでください。SEG-Y pipeline を optional feature に再設計する場合は別 issue にしてください。
